File: app/src/main/java/uk/co/sundroid/util/astro/BodyDay.java
package uk.co.sundroid.util.astro;
import java.util.Calendar;


public class BodyDay {
	
	private Calendar rise;
	
	private Calendar set;
	
	private Calendar transit;
	
	private double riseAzimuth;
	
	private double setAzimuth;
	
	private double transitAppElevation;
	
	private RiseSetType riseSetType;
	
	private double uptimeHours;

	public Calendar getRise() {
		return rise;
	}

	public void setRise(Calendar rise) {
		this.rise = rise;
	}

	public Calendar getSet() {
		return set;
	}

	public void setSet(Calendar set) {
		this.set = set;
	}

	public Calendar getTransit() {
		return transit;
	}

	public void setTransit(Calendar transit) {
		this.transit = transit;
	}

	public double getRiseAzimuth() {
		return riseAzimuth;
	}

	public void setRiseAzimuth(double riseAzimuth) {
		this.riseAzimuth = riseAzimuth;
	}

	public double getSetAzimuth() {
		return setAzimuth;
	}

	public void setSetAzimuth(double setAzimuth) {
		this.setAzimuth = setAzimuth;
	}

	public RiseSetType getRiseSetType() {
		return riseSetType;
	}

	public void setRiseSetType(RiseSetType riseSetType) {
		this.riseSetType = riseSetType;
	}

	public double getTransitAppElevation() {
		return transitAppElevation;
	}

	public void setTransitAppElevation(double transitAppElevation) {
		this.transitAppElevation = transitAppElevation;
	}

	public double getUptimeHours() {
		return uptimeHours;
	}

	public void setUptimeHours(double uptimeHours) {
		this.uptimeHours = uptimeHours;
	}
	
}

File: app/src/main/kotlin/uk/co/sundroid/util/astro/BodyDay.kt
package uk.co.sundroid.util.astro

import java.util.*
import uk.co.sundroid.util.astro.BodyDayEvent.Direction.*
import uk.co.sundroid.util.astro.BodyDayEvent.Event.*


open class BodyDay {

    @Deprecated("Prefer events")
    var rise: Calendar? = null

    @Deprecated("Prefer events")
    var set: Calendar? = null

    var events: MutableSet<BodyDayEvent> = TreeSet()

    @Deprecated("Prefer events")
    var transit: Calendar? = null

    @Deprecated("Prefer events")
    var riseAzimuth: Double = 0.0

    @Deprecated("Prefer events")
    var setAzimuth: Double = 0.0

    @Deprecated("Prefer events")
    var transitAppElevation: Double = 0.0

    var riseSetType: RiseSetType? = null

    var uptimeHours: Double = 0.0

    fun addEvent(event: BodyDayEvent) {
        events.add(event)
        when (event.direction) {
            RISING -> { if (eventUp[event.event] == null) eventUp[event.event] = event }
            DESCENDING -> { if (eventDown[event.event] == null) eventDown[event.event] = event }
        }
    }

    fun length(): Double {
        val rise = eventUp[RISESET]
        val set = eventDown[RISESET]
        if (rise != null && set != null) {
            return (set.time.timeInMillis - rise.time.timeInMillis) / (1000.0 * 60.0 * 60.0)
        } else if (rise == null && set == null && transitAppElevation < 0) {
            return 0.0
        }
        return 24.0
    }

    fun type(event: BodyDayEvent.Event): RiseSetType? {
        if (eventUp[event] == null && eventDown[event] == null) {
            return if (transitAppElevation > 0) {
                RiseSetType.RISEN
            } else {
                RiseSetType.SET
            }
        }
        return null
    }

    val eventUp: MutableMap<BodyDayEvent.Event, BodyDayEvent> = EnumMap(BodyDayEvent.Event::class.java)
    val eventDown: MutableMap<BodyDayEvent.Event, BodyDayEvent> = EnumMap(BodyDayEvent.Event::class.java)
    val eventType: MutableMap<BodyDayEvent.Event, RiseSetType> = EnumMap(BodyDayEvent.Event::class.java)

}
